crates/subspace-farmer/src/commitments/commitment_databases.rs

```diff
@@ -95,62 +103,125 @@ impl CommitmentDatabases {
             .collect()
     }
 
-    pub(super) fn get_db_entry(&self, salt: &Salt) -> Option<&Arc<DbEntry>> {
-        self.databases.peek(salt)
+    pub(super) fn get_db_entry(&self, salt: &Salt) -> Option<Arc<DbEntry>> {
+        return match salt {
+            _ if &self.first_db.read().as_ref().unwrap().salt == salt => {
+                self.first_db.read().clone()
+            }
+            _ if &self.second_db.read().as_ref().unwrap().salt == salt => {
+                self.second_db.read().clone()
+            }
+            _ => None,
+        };
     }
 
     /// Returns `Ok(None)` if entry for this salt already exists.
     pub(super) fn create_db_entry(
-        &mut self,
+        &self,
         salt: Salt,
+        current_salt: Salt,
     ) -> Result<Option<CreateDbEntryResult>, CommitmentError> {
-        if self.databases.contains(&salt) {
-            return Ok(None);
-        }
+        match salt {
+            _ if self.first_db.read().as_ref().is_some()
+                && self.first_db.read().as_ref().unwrap().salt == salt =>
+            {
+                return Ok(None)
+            }
+            _ if self.second_db.read().as_ref().is_some()
+                && self.second_db.read().as_ref().unwrap().salt == salt =>
+            {
+                return Ok(None)
+            }
+            _ => {} // don't do anything yet if salt does not exist
+        };
 
+        // create the new db_entry
         let db_entry = Arc::new(DbEntry {
             salt,
             db: Mutex::new(None),
         });
         let mut removed_entry_salt = None;
+        let old_db_entry;
 
-        let old_db_entry = if self.databases.len() >= COMMITMENTS_CACHE_SIZE {
-            self.databases.pop_lru().map(|(_salt, db_entry)| db_entry)
-        } else {
-            None
+        // delete the old database only, and don't touch the other to prevent data races
+        match current_salt {
+            // if the `first` DB is not the current one, replace it
+            _ if self.first_db.read().as_ref().is_none() => {
+                old_db_entry = self.first_db.write().replace(Arc::clone(&db_entry));
+            }
+            _ if self.second_db.read().as_ref().is_none() => {
+                old_db_entry = self.second_db.write().replace(Arc::clone(&db_entry));
+            }
+            _ if self.first_db.read().as_ref().is_some()
+                && self.first_db.read().as_ref().unwrap().salt != current_salt =>
+            {
+                old_db_entry = self.first_db.write().replace(Arc::clone(&db_entry));
+            }
+            // if the `second` DB is not the current one, replace it
+            _ if self.second_db.read().as_ref().is_some()
+                && self.second_db.read().as_ref().unwrap().salt != current_salt =>
+            {
+                old_db_entry = self.second_db.write().replace(Arc::clone(&db_entry));
+            }
+            _ => unreachable!(),
```

Comment:
I'm not going to analyze deeper for now because this is already problematic. You check with `.read()`, release the lock and then do `.write()`, opening window for race conditions where something can modify contents between `.read()` and `.write()` calls.

Reply:
I don't think it is problematic. Yes, something else can modify the contents in between the read/write locks. But it won't. AFAIK, nothing is modifying the inactive commitments database, it is already the old one. So basically, this race condition will never happen in our protocol (unless there is a detail I'm missing).

Reply:
Not only inactive, active database has the same pattern here. It is problematic because the way API is written right now is racy and if it is racy, then there will likely be a bug down the line that will hit this and make commitments databases inconsistent and will be very hard to diagnose. So the code should be such that it doesn't have race conditions.

Also on a general note:
```rust
self.second_db.read().as_ref().is_some() && self.second_db.read().as_ref().unwrap()
```

^ is anti-pattern, use `if let` instead.

Also FYI `RwLock` is slower than `Mutex` and is only rarely useful for cases where there are many readers that need to read things concurrently. It is far less useful that it seems initially.

Reply:
> Not only inactive, active database has the same pattern here
I have to disagree. In this pattern here, we are not acquiring the write lock for the active database. 

> RwLock is only rarely useful for cases where there are many readers that need to read things concurrently
I agree, and that is what I'm trying to achieve. To be able to detect which database to overwrite, we have to read both databases to see which is the old one, and at the same time, another thread will be reading the active database for finding solutions. Seems to me like multiple readers, single writer problem.

Reply:
> we are not acquiring the write lock for the active database.

What is `old_db_entry = self.first_db.write().replace(Arc::clone(&db_entry));` then?

> To be able to detect which database to overwrite, we have to read both databases to see which is the old one, and at the same time, another thread will be reading the active database for finding solutions.

That might need additional data structure or other changes to be implemented without locking active database. I'll have to take a closer look at this, but I definitely can't approve above code where there is a chance for race conditions on both active and inactive databases. It may even panic if between `.is_some()` and `.unwrap()` underlying data structure changes.

Reply:
> What is old_db_entry = self.first_db.write().replace(Arc::clone(&db_entry)); then?

`first` and `second` are just names.  In one turn, `first` will be the old, and in the next one, `second` will be the old, this will continue forever.